aggregator/election_mgr.go

```diff
@@ -0,0 +1,545 @@
+	if err != nil {
+		return err
+	}
+	go mgr.watchGoalStateChanges(stateChangeWatch)
+
+	_, verifyWatch, err := mgr.goalStateWatchable.Watch()
+	if err != nil {
+		return err
+	}
+	go mgr.verifyPendingFollower(verifyWatch)
+
+	go mgr.campaignLoop()
+	go mgr.reportMetrics()
+	return nil
+}
+
+func (mgr *electionManager) ElectionState() ElectionState {
+	return mgr.electionStateWatchable.Get().(ElectionState)
+}
+
+func (mgr *electionManager) Resign(ctx context.Context) error {
+	_, watch, err := mgr.electionStateWatchable.Watch()
+	if err != nil {
+		return fmt.Errorf("error creating watch when resigning: %v", err)
+	}
+	defer watch.Close()
+
+	resignAction := resignAction{errCh: make(chan error, 1)}
+	select {
+	case mgr.resignCh <- resignAction:
+	default:
+		return errResignActionAlreadyEnqueued
+	}
+
+	for {
+		select {
+		case <-watch.C():
+			if state := watch.Get().(ElectionState); state == FollowerState {
+				return nil
+			}
+		case <-ctx.Done():
+			mgr.metrics.resignTimeout.Inc(1)
+			mgr.logError("resign error", ctx.Err())
+			return ctx.Err()
+		case err := <-resignAction.errCh:
+			mgr.metrics.resignErrors.Inc(1)
+			mgr.logError("resign error", err)
+			return err
+		}
+	}
+}
+
+func (mgr *electionManager) Close() error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.state != electionManagerOpen {
+		return errElectionManagerNotOpenOrClosed
+	}
+	close(mgr.doneCh)
+	mgr.state = electionManagerClosed
+	return nil
+}
+
+func (mgr *electionManager) watchGoalStateChanges(watch xwatch.Watch) {
+	defer watch.Close()
+
+	for {
+		select {
+		case <-mgr.doneCh:
+			return
+		case <-watch.C():
+			mgr.processGoalState(watch.Get().(goalState))
+		case action := <-mgr.resignCh:
+			mgr.resign(action)
+		}
+	}
+}
+
+func (mgr *electionManager) processGoalState(goalState goalState) {
+	currState := mgr.ElectionState()
+	newState := goalState.state
+	if currState == newState {
+		return
+	}
+	// NB(xichen): the follower to pending follower transition is conceptually invalid
+	// but because the leader election API currently does not verify states, if the client
+	// receives a follower state, it can only treat it as a "pendingFollower" state. This
+	// is no longer needed once the leader election logic is adapted to only notify clients
+	// of states that have been verified.
+	if currState == FollowerState && newState == PendingFollowerState {
+		return
+	}
+	mgr.electionStateWatchable.Update(newState)
+	mgr.metrics.stateChanges.Inc(1)
+	mgr.logger.Info(fmt.Sprintf("election state changed from %v to %v", currState, newState))
+}
+
+func (mgr *electionManager) verifyPendingFollower(watch xwatch.Watch) {
+	defer watch.Close()
+
+	for {
+		select {
+		case <-mgr.doneCh:
+			return
+		case <-watch.C():
+		}
+
+		currState := mgr.goalStateWatchable.Get().(goalState)
+		if currState.state != PendingFollowerState {
+			continue
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), mgr.changeVerifyTimeout)
```

Comment:
We do, although this is a golang context and the standard way of closing it is to call `cancel` down below at line 402.